### ISSUE TITLE
Parser fails when union subquery contains correlated subquery

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#11235]
The parser currently fails parsing this perfectly valid query:

```sql
select 1, 2 where 1 = (select 1) 
union all 
select 2, 1 where 2 = (select 2)
```

The exception being:

```
org.jooq.impl.ParserException: Select list must contain 1 columns. Got: 2: [1:56] ...t 1, 2 where 1 = (select 1) union all select 2, 1 [*]where 2 = (select 2)
	at org.jooq.impl.ParserContext.exception(ParserImpl.java:12349)
	at org.jooq.impl.ParserContext.parseQueryPrimary(ParserImpl.java:1511)
	at org.jooq.impl.ParserContext.parseQueryTerm(ParserImpl.java:1408)
	at org.jooq.impl.ParserContext.parseQueryExpressionBody(ParserImpl.java:1385)
	at org.jooq.impl.ParserContext.parseSelect(ParserImpl.java:1154)
	at org.jooq.impl.ParserContext.parseSelect(ParserImpl.java:1149)
	at org.jooq.impl.ParserContext.parseTerm(ParserImpl.java:7277)
	at org.jooq.impl.ParserContext.parseUnaryOps(ParserImpl.java:6555)
	at org.jooq.impl.ParserContext.parseExp(ParserImpl.java:6532)
	at org.jooq.impl.ParserContext.parseFactor(ParserImpl.java:6509)
	at org.jooq.impl.ParserContext.parseSum(ParserImpl.java:6455)
	at org.jooq.impl.ParserContext.parseNumericOp(ParserImpl.java:6440)
	at org.jooq.impl.ParserContext.parseCollated(ParserImpl.java:6414)
	at org.jooq.impl.ParserContext.parseConcat(ParserImpl.java:6404)
	at org.jooq.impl.ParserContext.parsePredicate(ParserImpl.java:5392)
	at org.jooq.impl.ParserContext.parseNot(ParserImpl.java:5275)
	at org.jooq.impl.ParserContext.parseAnd(ParserImpl.java:5265)
	at org.jooq.impl.ParserContext.parseOr(ParserImpl.java:5256)
	at org.jooq.impl.ParserContext.parseCondition(ParserImpl.java:5252)
	at org.jooq.impl.ParserContext.parseQueryPrimary(ParserImpl.java:1554)
	at org.jooq.impl.ParserContext.parseQueryTerm(ParserImpl.java:1408)
	at org.jooq.impl.ParserContext.parseQueryExpressionBody(ParserImpl.java:1375)
	at org.jooq.impl.ParserContext.parseSelect(ParserImpl.java:1154)
	at org.jooq.impl.ParserContext.parseSelect(ParserImpl.java:1149)
	at org.jooq.impl.ParserContext.parseQuery(ParserImpl.java:1006)
	at org.jooq.impl.ParserContext.parseQuery0(ParserImpl.java:784)
	at org.jooq.impl.ParserImpl.parseQuery(ParserImpl.java:620)
	at org.jooq.impl.ParserImpl.parseQuery(ParserImpl.java:615)
	at org.jooq.impl.AbstractParserTest.lambda$3(AbstractParserTest.java:323)
	at org.jooq.impl.AbstractParserTest.assertEquality(AbstractParserTest.java:417)
	at org.jooq.impl.AbstractParserTest.assertQueryEquals(AbstractParserTest.java:323)
	at org.jooq.impl.ParserImplTest.testSelectUnionWithCorrelatedSubqueries(ParserImplTest.java:792)
```

**Comment by lukaseder** [comment]
The problem is with the parser thinking that the union operator belongs to the correlated subquery

**Comment by lukaseder** [comment]
A workaround would be to parse this instead:

```sql
select 1, 2 where 1 = (select 1) and 1 = 1 
union all 
select 2, 1 where 2 = (select 2)
```

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.15.0 and 3.14.5 (#11236)
